various ports: remove subports/variants for EOL Python versions

reneeotten opened the pull request:
#### Description
This PR removes subports and/or variants for EOL Python versions. The (large) set of ports is internally consistent so that (hopefully) there are no ports left that have a dependency on a removed subport.

No other changes (except for removal of redundant fields) so if a port doesn't build anymore it is (or at least should) not caused by this PR.

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 10.15.7 19H1824 x86_64
Xcode 12.4 12D4e

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

Schamschula commented:
I'm glad to see this! I hadn't deleted these subports myself, as they still had dependents.

chrstphrchvz commented:
Notifying maintainers not notified by @macportsbot (https://trac.macports.org/ticket/63821):

@blair for port py-tensorflow_estimator1.
@dfm for port py-corner.
@smithsp for port py-fortranformat.
@mojca for port py-django-countries, py-paho-mqtt, py-pycson, py-pypeg2, py-sounddevice, py-spacepy, py-speg.
@zdw for port py-dockerpy-creds.
@essandess for port py-graph-tool.
@jswhit for port py-pygrib.
@SoapZA for port py-pysam.
@ierofant for port py-bitstruct, py-canmatrix.
@donkopotamus for port py-dispatcher.

jjstickel commented:
LGTM! Thanks for doing this.

smithsp commented on python/py-fortranformat/Portfile:
```diff
@@ -20,11 +19,10 @@ long_description    Generates text from a Python list of variables or will \
     statement passed.
 
-python.versions     27 35 36 37 38
+python.versions     37 38
```
I approve.

smithsp reviewed (approved):
I approve.

kurthindenburg commented:
If you've verified no other ports need them, go ahead.  I know a lot of mine had a lot of other ports using them.

xeron commented:
There're some ports outside of python group which depend on these older version, they may need a cleanup too, e.g.:

https://github.com/macports/macports-ports/blob/d275a24edbefe7f361f2aa6eb8f59648092058c0/databases/pgcli/Portfile#L24-L25
https://github.com/macports/macports-ports/blob/d275a24edbefe7f361f2aa6eb8f59648092058c0/databases/mycli/Portfile#L24-L25

cjones051073 commented:
This looks good to me.

I also vote for merging quick and if deps from variants etc. then pop on address them later on. Merging this is the easiest way to find them ;)

blair commented:
I approve (I don't see an approve button where).

michaelld commented on python/py-construct/Portfile:
```diff
@@ -32,7 +31,7 @@ homepage            https://construct.readthedocs.org
 if {${name} ne ${subport}} {
 
-    if { ${python.version} == 27 || ${python.version} == 35} {
+    if { ${python.version} == 27} {
```
Should this be removed?

reneeotten replied:
no, the py27 subport will need to stay at least for a while since it has a dependent:
```
% port echo depends:py27-construct
gr37-satellites
```

michaelld replied:
Gotcha. Somehow I hadn't made the connection that you're removing Py27 ports with no dependencies ... makes sense now!

michaelld reviewed (commented):
I am ambivalent about removing support for Py27. I have no objection to removing Py36 & TBH even Py37 if we want to. What about `py-scipy` and `py-numpy` And also the port I comment on here with respect to Py27 removal?

cjones051073 commented:
> I approve (I don't see an approve button in the UI).

View the changes, then Start a review and just submit your approval there with no comments.

reneeotten commented:
> I am ambivalent about removing support for Py27. I have no objection to removing Py36 & TBH even Py37 if we want to. What about `py-scipy` and `py-numpy` And also the port I comment on here with respect to Py27 removal?

in the long run, yes these should/will be removed - but one has to start with the ones that have no dependents and work towards the ones higher in the food-chain like `py-numpy` and `py-scipy`.

mascguy commented:
Renee, you might want to use `[skip ci]` in the commit message, as the CI jobs are highly unlikely to complete successfully. (And the resources will also be tied up for six hours.)

reneeotten commented:
> Renee, you might want to use `[skip ci]` in the commit message, as the CI jobs are highly unlikely to complete successfully. (And the resources will also be tied up for six hours.)

sure, but where? Adding it for one random commit in the whole list will suffice, or first, last, ....?

mascguy commented:
> > Renee, you might want to use `[skip ci]` in the commit message, as the CI jobs are highly unlikely to complete successfully. (And the resources will also be tied up for six hours.)
> 
> sure, but where? Adding it for one random commit in the whole list will suffice, or first, last, ....?

Good point, might be better to simply cancel CI manually then!  :-)

michaelld reviewed (approved):
LGTM & makes sense to remove Py27 ports with no dependencies!

lbschenkel commented:
I am OK with whatever the consensus ends up being.

i0ntempest commented:
No objections here

nerdling commented:
> @jmroot Josh, do you have a TCL script to do a full exhaustive search of dependent ports, including all variants? (It's a bit of a pain for dependents of Python ports, given the use of `${python.version}` and such within names. Making a simple grep less-than-conclusive...)

Would `grep -r --include=Portfile py.*psycopg2` suffice?

reneeotten commented:
> Well, in the interest of not causing pain for our users and maintainers, I think we need to do more exhaustive research before merging this. Particularly given that this isn't time-critical.

Frankly, I have done as much due diligence on this as I am willing/able to do given the amount of time it takes. If maintainers specifically object to the removal of Python versions for "their" ports I will take them out. If not, I intend to merge this PR after the usual three days waiting period. If anyone wants to go ahead and search the tree for occurrences in variants for the ~270 ports in this PR and list them here, please feel free - the clock is ticking ;) No seriously, I am fine with waiting a bit longer to merge this if someone is committing to do that work - but given how these exercises happened in the past, I don't hold my breath. 

Of course, I am more than happy to fix any possible, minor fallout that this could cause *after* the fact; I tried to be careful and not break anything with this PR similarly as I've done in the past.

mascguy commented:
> > Well, in the interest of not causing pain for our users and maintainers, I think we need to do more exhaustive research before merging this. Particularly given that this isn't time-critical.
> 
> Frankly, I have done as much due diligence on this as I am willing/able to do given the amount of time it takes.

Well, that's why I added @jmroot to the discussion: There might be a TCL script to handle this, and/or built-in functionality in base.

So you needn't spend more time on it Renee, as we're offering to help. Just don't merge until others have had a chance to help you proactively vet the dependencies!  ;-)

JacksonIsaac reviewed (approved):
lg2m for the ports I am in maintainer list

reneeotten commented:
I have rebased with `master` and fixed the conflicts; not sure why the GitHub website still thinks that it cannot be rebased... 

I'll give this another day so that people had more than a week to respond and then will merge this PR provided that I can figure out how to resolve this, or commit directly if that works out-of-the-box.

mascguy commented:
> @jmroot Josh, do you have a TCL script to do a full exhaustive search of dependent ports, including all variants? (It's a bit of a pain for dependents of Python ports, given the use of `${python.version}` and such within names. Making a simple grep less-than-conclusive...)

@jmroot, do you have a TCL script to run an exhaustive dependency search like this, across all variants?

lpsinger commented on python/py-emcee/Portfile:
```diff
@@ -17,23 +17,17 @@ long_description    emcee is a stable, well tested Python implementation of \
 
 if {${name} ne ${subport}} {
-    if {${python.version} >= 36} {
-        conflicts   py${python.version}-emcee3
```
Since we have dropped support for Python <= 3.6, the condition in the `if` statement is always true, so the `conflicts` statement needs to be kept.

reneeotten replied:
good catch, I missed that one - fixed locally!

lpsinger replied:
It's not showing up in GitHub yet...

reneeotten replied:
correct, as I said “fixed locally” and I haven’t pushed yet to this PR

lpsinger commented on python/py-gcn/Portfile:
```diff
@@ -6,10 +6,9 @@ PortGroup           github 1.0
 github.setup        lpsinger pygcn 0.1.19 v
 name                py-gcn
-python.versions     27 35 36 37
+python.versions     37
```
```suggestion
python.versions     37 38 39 310
```

reneeotten replied:
while I in principle agree with adding subports, I will not do so in this PR which is purely about removing EOL versions. Adding newer Python subports requires one to make sure upstream supports them and such, that's something for a maintainer to do at another time.

lpsinger commented on python/py-gcn/Portfile:
```diff
@@ -6,10 +6,9 @@ PortGroup           github 1.0
 github.setup        lpsinger pygcn 0.1.19 v
 name                py-gcn
-python.versions     27 35 36 37
+python.versions     37
```
```suggestion
python.versions     37 38 39 310
```

mascguy commented:
Rene, I'm good with merging as well. As you said before, it's easy enough to fix anything broken after-the-fact.

mascguy commented:
Renee, it looks like port `volatility` is now broken, due to the removal of `py27-distorm`:

```
$ port deps volatility
Full Name: volatility @2.5_1
Build Dependencies:   py27-setuptools
Library Dependencies: py27-pycryptodome, py27-distorm, python27
Runtime Dependencies: yara

$ port rdeps volatility
Error: Port py27-distorm not found
```

So either `volatility` needs to be updated to use a later version of `py-distorm`, or the Python 2.7 version needs to be restored.

mascguy commented:
> So either `volatility` needs to be updated to use a later version of `py-distorm`, or the Python 2.7 version needs to be restored.

It looks like `py-distorm` was removed entirely, which I didn't expect. So that ports needs to be restored ASAP.

nerdling commented:
Or we can get rid of `volatility`.

mascguy commented:
> Or we can get rid of `volatility`.

While I can't speak to how widely it's used by our users, I do use it personally. So I'm going to veto that idea!  ;-)

jmroot commented:
There seems to have been some confusion between `distnorm` and `distorm`. The commit restores the correct name, but it looks like the other changes should be kept?

mascguy commented:
> There seems to have been some confusion between `distnorm` and `distorm`. The commit restores the correct name, but it looks like the other changes should be kept?

The correct name appears to be `distorm`, per the project site: https://github.com/gdabah/distorm

`distnorm` appears to be a typo. The portfiles are also essentially identical, so `distnorm` needs to be removed. And dependents, if any, need to be updated to reference `py-distorm`

jmroot commented:
Yes, as I said, the commit restored the correct name but also reverted some other changes.

mascguy commented:
> Yes, as I said, the commit restored the correct name but also reverted some other changes.

My commit only involved resurrection of `py-distorm`. What "other changes" are you referring to?

mascguy commented:
> > > Yes, as I said, the commit restored the correct name but also reverted some other changes.
> > 
> > 
> > My commit only involved resurrection of `py-distorm`. What "other changes" are you referring to?
> 
> Looks to me like it did more
> 
> [20fc94e](https://github.com/macports/macports-ports/commit/20fc94e702fa3ba8fa88d9f7e8603525f5752725)
> 
> you added back the 35 and 36 supports. I've since removed them again.

Correct, because I restored the port, based on the state prior to Renee's changes. In particular, the inadvertent renaming of the port to `py-distnorm`.

mascguy commented:
Renee, I don't have the time to review the commits in this PR, given the sheer number of changes.

Can you please audit this work, and verify that no other ports were inadvertently renamed?

mascguy commented:
> > Correct, because I restored the port, based on the state prior to Renee's changes. In particular, the inadvertent renaming of the port to py-distnorm.
> 
> And we're saying only the accidental name change should have been reverted, not the other, intentional, changes.

Yes, that's well-understood, and noted.

However, my primary concern is ensuring we don't have a broken ports tree.

jmroot commented:
> Can you please audit this work, and verify that no other ports were inadvertently renamed?

Doesn't look like it. But this is a good reminder of why changes that aren't strictly related should be avoided in mass commits like this.

mascguy commented:
> > However, my primary concern is ensuring we don't have a broken ports tree.
> 
> Well, that has yet to be accomplished in the general sense, so good luck. :)

It's a labor of love, but I'm trying!  ;-)

reneeotten commented:
oh well.... sorry about that! Just changing the `name` field to correct the typo though would have fixed all of this - no need for mass panic ;)

mascguy commented:
> oh well.... sorry about that! Just changing the `name` field to correct the typo though would have fixed all of this - no need for mass panic ;)

Yeah, sorry for being a bit dramatic folks!  :-D

adfernandes commented:
Thanks for this, @reneeotten!
